src/cache.rs

```diff
@@ -0,0 +1,292 @@
+                                            sqlite: &'sqlite rusqlite::Connection,
+                                            schema: &'schema Schema) -> Result<()> {
+        // fetch results and add to cache
+        self.cache = self.value_provider.fetch_values(sqlite, schema)?;
+        Ok(())
+    }
+
+    fn update_values<'sqlite, 'schema>(&mut self,
+                                      sqlite: &'sqlite rusqlite::Connection,
+                                      schema: &'schema Schema) -> Result<()> {
+        self.cache_values(sqlite, schema)
+    }
+
+    fn get(&self, key: &Self::Key) -> Option<&Self::Value> {
+        self.cache.get(&key)
+    }
+}
+
+pub struct AttributeCacher {
+    cache: BTreeMap<NamespacedKeyword, EagerCache<TypedValue, Vec<TypedValue>, AttributeValueProvider>>,   // values keyed by attribute
+}
+
+impl AttributeCacher {
+
+    pub fn new() -> Self {
+        AttributeCacher {
+            cache: BTreeMap::new(),
+        }
+    }
+
+    pub fn register_attribute<'sqlite, 'schema>(&mut self, sqlite: &'sqlite rusqlite::Connection, schema: &'schema Schema, attribute: NamespacedKeyword) -> Result<()> {
+        if schema.identifies_attribute(&attribute) {
+            let value_provider = AttributeValueProvider{ attribute: attribute.clone() };
+            let mut cacher = EagerCache::new(value_provider);
+            cacher.cache_values(sqlite, schema)?;
+            self.cache.insert(attribute, cacher);
+            Ok(())
+        } else {
+            bail!(ErrorKind::UnknownAttribute(attribute.to_string()))
+        }
+    }
+
+    pub fn deregister_attribute(&mut self, attribute: &NamespacedKeyword) -> Option<BTreeMap<TypedValue, Vec<TypedValue>>> {
+        self.cache.remove(attribute).map(|m| m.cache )
+    }
+
+    pub fn get(&mut self, attribute: &NamespacedKeyword) -> Option<&BTreeMap<TypedValue, Vec<TypedValue>>> {
+        self.cache.get( attribute ).map(|m| &m.cache )
+    }
+
+    pub fn get_for_entid(&mut self, attribute: NamespacedKeyword, entid: KnownEntid) -> Option<&Vec<TypedValue>> {
+        if let Some(c) = self.cache.get(&attribute) {
+            c.get(&entid.into())
+        } else { None }
+    }
+}
+
+struct AttributeValueProvider {
+    attribute: NamespacedKeyword,
+}
+
+impl ValueProvider<TypedValue, Vec<TypedValue>> for AttributeValueProvider {
+    fn fetch_values<'sqlite, 'schema>(&mut self, sqlite: &'sqlite rusqlite::Connection, schema: &'schema Schema) -> Result<BTreeMap<TypedValue, Vec<TypedValue>>> {
+        Ok(q_once(sqlite,
+                  schema,
+                  r#"[:find ?entity ?value
+                                          :in ?attribute
+                                          :where [?entity ?a        ?value]
+                                                 [?a      :db/ident ?attribute]]"#,
+                  QueryInputs::with_value_sequence(vec![(Variable::from_valid_name("?attribute"), self.attribute.clone().into())]))
```

Comment:
You have the schema at this point. Use `HasSchema::attribute_for_ident` earlier to strip away the keyword down to a known attribute's entid. No point in joining against `:db/ident`!